src/staking/index.ts

```diff
@@ -0,0 +1,170 @@
+        org: string,
+        /** minimum staking period in seconds */
+        minStakingPeriod: number | utils.BigNumber,
+        /** patron's part of the reward in fractions of thousandth */
+        patronRewardPortion: number,
+        /** roles required to stake */
+        patronRoles: string[],
+        /** stake put by service provider when pool is launched */
+        principal: utils.BigNumber
+      }
+  ): Promise<void> {
+    (await this._stakingPoolFactory.launchStakingPool(
+      namehash(org),
+      minStakingPeriod,
+      patronRewardPortion,
+      patronRoles.map((r) => namehash(r))
+      ,
+      { value: principal.toHexString() }
+    )).wait();
+  }
+
+  /**
+   * @description Returns all services for which pools are launched
+   */
+  async allServices(): Promise<Service[]> {
+    const orgs = await this._stakingPoolFactory.orgsList();
+    return Promise.all(
+      orgs.map(
+        (org) => this._stakingPoolFactory.services(org)
+          .then((service) => ({ ...service, org }))
+          .then((service) => this._domainReader.readName(service.org)
+            .then((org) => ({ ...service, org }))
+          )));
+  }
+
+  /**
+   * @description Returns pool launched for `org` if any
+   * @param org ENS name of organization
+   */
+  async getPool(org: string): Promise<StakingPool | null> {
+    const service = await this._stakingPoolFactory.services(namehash(org));
+    if (service.pool === emptyAddress) {
+      return null;
+    }
+    const pool = new StakingPool__factory(this._signer).attach(service.pool);
+    return new StakingPool(pool);
+  }
+}
+
+/**
+ * Abstraction over staking pool smart contract
+ */
+export class StakingPool {
+  constructor(private pool: StakingPoolContract) { }
+
+  /**
+   * @description Locks stake and starts accumulating reward
+   * @emits StakingPool.StakePut
+   */
+  async putStake(
+    /** stake amount */
+    stake: utils.BigNumber,
+  ): Promise<void> {
+    (await this.pool.putStake({
+      value: stake.toHexString()
+    })).wait();
+  }
+
+  /**
+   * Accumulated reward
+   */
+  async checkReward(): Promise<utils.BigNumber> {
+    return this.pool.checkReward();
+  }
+
+  /**
+   * @param patron staker address
+   * @returns Stake
+   */
+  async getStake(patron?: string): Promise<Stake> {
+    if (!patron) {
+      patron = await this.pool.signer.getAddress();
+    }
+    return this.pool.stakes(patron);
+  }
+
+  /**
+  * @description Stops accumulating of the reward and prepars stake to withdraw after withdraw delay. 
+  * Withdraw request unavailable until minimum staking period ends
+  */
+  async requestWithdraw(): Promise<void> {
+    (await this.pool.requestWithdraw()).wait();
+  }
+
+  /**
+   * @description pays back stake with accumulated reward. Withdrawn unavailable until withdrawn delay ends
+   * @emits StakingPool.StakeWithdrawn
+   */
+  async withdraw(): Promise<void> {
+    (await this.pool.withdraw()).wait();
+  }
```

Comment:
Might be nice to have a `canWithdraw()` method. Otherwise, I think the only way for a client to, for example, disable a withdraw button until they can withdraw, is to attempt the `withdraw()` and catch an exception. But maybe @DawidGil-jlabs you can indicate `canWithdraw()` would be helpful or not.

Reply:
Good catch. I need to disable somehow withdraw button. So if that's not a problem for you @JGiter, please add method `canWithdraw()`. I can create method that will ask for canWithdraw every few seconds so we will have actual information on the screen.